greenlight/harness/testcase.py

```diff
@@ -19,10 +19,17 @@ def setUp(self, *args, **kwargs):
         MarionetteTestCase.setUp(self, *args, **kwargs)
         Puppeteer.set_marionette(self, self.marionette)
 
+        self._start_handle_count = len(self.marionette.window_handles)
         self.marionette.set_context('chrome')
         self.browser = self.windows.current
 
     def tearDown(self, *args, **kwargs):
-        MarionetteTestCase.tearDown(self, *args, **kwargs)
+        win_count = len(self.marionette.window_handles)
+        self.assertEqual(win_count, self._start_handle_count,
+                         "A test must not leak window handles. "
+                         "This test started the browser with %s open "
+                         "top level browsing contexts, but ended with %s." %
+                         (self._start_handle_count, win_count))
 
         self.prefs.restore_all_prefs()
+        MarionetteTestCase.tearDown(self, *args, **kwargs)
```

Comment:
I think things could get worse if we allow tests to leak windows. They could easily do things like leave marionette switched into the wrong handle and cause future tests to fail. You mean we're going to have a problem because subsequent tear down code will not run in case of a failure of this assertion? I think that might be ok -- later tests might fail, but the problem will be fairly apparent given the output, and usually straightforward to fix.